hail/src/main/scala/is/hail/expr/ir/InferPType.scala

```diff
@@ -0,0 +1,460 @@
+// FIXME: strip all requiredness logic when possible
+// TODO: We don't implicitly walk bottom-up for virtualTypes (not calling .typ) in all cases
+// TODO: If virtual type inference doesn't occur before this step, there may be subtle
+// TODO: behavior differences between IR Nodes (those that need .typ here and those that don't)
+// TODO: In cases where env is sent as is to child nodes, do we need copy? I believe no, as long as copy-on-write
+
+// TODO: aggregators: 1) use bindAgg/bindScan instead of bindEval? 2) proper passing of env, seems little needed
+object InferPType {
+  def apply(ir: IR, env: BindingEnv[PType]): PType = {
+    ir match {
+      case I32(_) => PInt32()
+      case I64(_) => PInt64()
+      case F32(_) => PFloat32()
+      case F64(_) => PFloat64()
+      case Str(_) => PString()
+      // TODO: The following (where canonical) have no children
+      // TODO: OK to be canonical for now? Later we will want to
+      // TODO: use properties of env?
+      case Literal(t, _) => PType.canonical(t)
+      case True() | False() => PBoolean()
+      case Void() => PVoid
+      case Cast(_, t) => PType.canonical(t)
+      case CastRename(_, t) => PType.canonical(t)
+      case NA(t) => PType.canonical(t)
+      case IsNA(_) => PBoolean()
+      case Coalesce(values) => {
+        val vit = values.iterator
+        val head = values.iterator.next()
+        head.inferSetPType(env)
+
+        while(vit.hasNext) {
+          val value = vit.next()
+          value.inferSetPType(env)
+
+          assert(head.pType == value.pType)
+        }
+      }
+      // TODO: Why is there a type (var _type) on Ref? Cache to avoid lookup? if so seems area for bugs
+      case Ref(name, _) => {
+        env.eval.lookup(name)
+      }
+      case In(_, t) => PType.canonical(t)
+      // TODO: need to descend into args?
+      case MakeArray(_, t) => PType.canonical(t)
+      case MakeStream(_, t) => PType.canonical(t)
+      case MakeNDArray(data, shape, _) => {
+        data.inferSetPType(env)
+        shape.inferSetPType(env)
+
+        // TODO: Is this wrong?
+        val nElem = shape.pType.asInstanceOf[PTuple].size
+
+        // TODO: requiredeness?
+        PNDArray(coerce[PArray](data.pType).elementType, nElem)
+      }
+
+      case _: ArrayLen => PInt32()
+      case _: ArrayRange => PArray(PInt32())
+      case _: StreamRange => PStream(PInt32())
+      case _: LowerBoundOnOrderedCollection => PInt32()
+      case _: ArrayFor => PVoid
+      case _: InitOp => PVoid
+      case _: SeqOp => PVoid
+      case _: Begin => PVoid
+      // FIXME in IR suggests Die should give type any, fix this once that is fixed
+      // but in test, TFloat64() is used, which will fit
+      case Die(_, t) => PType.canonical(t)
+      case If(cond, cnsq, altr) => {
+        cond.inferSetPType(env)
+        cnsq.inferSetPType(env)
+        altr.inferSetPType(env)
+
+        assert(cond.pType == PBoolean())
+
+        if (cnsq.pType != altr.pType)
```

Comment:
When we have real non-canonical ptypes we'd either want to use a union type, or coerce one ptype to another.

Reply:
good for now though